lib/model-builder.js

```diff
@@ -388,9 +411,58 @@ ModelBuilder.prototype.define = function defineClass(className, properties, sett
       }
     }
 
-    // Merge the settings
+    // Merging the settings
     var originalSubclassSettings = subclassSettings;
-    subclassSettings = mergeSettings(settings, subclassSettings);
+    if (!true) { // TODO: define a feature flag value
+      // legacy behaviour
+      subclassSettings = mergeSettings(settings, subclassSettings);
+    } else {
+      // new behaviour
+      let mergePolicy = ModelClass.mergePolicy;
+      // deep clone settings
+      let mergedSettings = mergeSettings(settings);
+
+      Object.keys(settings).forEach(function(key) {
+        // rank base class settings arrays elements where required
+        if (mergePolicy[key] && mergePolicy[key].rank) {
+          settings[key] = rankArrayElements(settings[key], ModelBaseClass.rank);
+        }
+      });
+
+      Object.keys(subclassSettings).forEach(function(key) {
+        // assign default merge policy to unknown settings properties
+        mergePolicy[key] = mergePolicy[key] || mergePolicy.default;
+
+        // allow null value to remove unwanted properties from base class settings
+        if (subclassSettings[key] === null) {
+          delete mergedSettings[key];
+          return;
+        }
+        // rank sub class settings arrays elements where required
+        if (mergePolicy[key].rank) {
+          subclassSettings[key] = rankArrayElements(subclassSettings[key], ModelBaseClass.rank + 1);
+        }
+        // replace base class settings where required
+        if (mergePolicy[key].replace) {
+          mergedSettings[key] = subclassSettings[key];
+          return;
+        }
+        // patch inner properties of base class settings where required
+        if (mergePolicy[key].patch) {
+          let mergedSetting = mergedSettings[key];
+          let subclassSetting = subclassSettings[key];
+          Object.keys.forEach(function(k) {
+            mergedSetting[k] = subclassSetting[k];
+          });
+          return;
+        }
+
+        // in case no merge policy matched, apply a deep merge
+        mergedSettings[key] = mergeSettings(settings[key], subclassSettings[key]);
+      });
+      // assign final sub class settings
+      subclassSettings = mergedSettings;
+    }
```

Comment:
of course this can be moved back to a separate function or separate file later on in this PR.

👉  the main point at this stage is that it leaves the original mergeSettings function untouched, and relies on it for legitimate deep merge operations